frontend/device/generic/device.lua

```diff
@@ -144,6 +144,31 @@ function Device:onPowerEvent(ev)
         self.screen_saver_mode = false
         self.powerd:refreshCapacity()
         self.powerd:afterResume()
+    elseif ev == "PowerLong" and not self.screen_saver_mode then
+        local UIManager = require("ui/uimanager")
+        local ConfirmBox = require("ui/widget/confirmbox")
+        local InfoMessage = require("ui/widget/infomessage")
+        local ReaderUI = require("apps/reader/readerui")
+        local FileManager = require("apps/filemanager/filemanager")
```

Comment:
@chrox That's what I've tried first, but it seems to have no effect, i.e. without the `os.execute("poweroff")` koreader just continue executing and it doesn't save the state of the open document.